lib/decompress/zstd_decompress_block.c

```diff
@@ -121,8 +129,32 @@ size_t ZSTD_decodeLiteralsBlock(ZSTD_DCtx* dctx,
                     litCSize = (lhc >> 22) + ((size_t)istart[4] << 10);
                     break;
                 }
+                RETURN_ERROR_IF(litSize > 0 && dst == NULL, dstSize_tooSmall, "NULL not handled");
                 RETURN_ERROR_IF(litSize > ZSTD_BLOCKSIZE_MAX, corruption_detected, "");
                 RETURN_ERROR_IF(litCSize + lhSize > srcSize, corruption_detected, "");
+                RETURN_ERROR_IF(expectedWriteSize < litSize , dstSize_tooSmall, "");
+                if (streaming == not_streaming && dstCapacity > ZSTD_BLOCKSIZE_MAX + WILDCOPY_OVERLENGTH + litSize + WILDCOPY_OVERLENGTH)
+                {
+                    /* room for litbuffer to fit after output without read faulting */
+                    dctx->litBuffer = (BYTE*)dst + ZSTD_BLOCKSIZE_MAX + WILDCOPY_OVERLENGTH;
+                    dctx->litBufferEnd = dctx->litBuffer + litSize;
+                    dctx->splitLitBuffer = 0;
+                }
+                else if (litSize > ZSTD_LITBUFFEREXTRASIZE)
+                {
+                    /* won't fit in litExtraBuffer, so it will be split between end of dst and extra buffer */
+                    /* initially this will be stored entirely in dst during huffman decoding, it will partially shifted to litExtraBuffer after */
+                    dctx->litBuffer = (BYTE*)dst + expectedWriteSize - litSize; 
+                    dctx->litBufferEnd = (BYTE*)dst + expectedWriteSize;
+                    dctx->splitLitBuffer = 1;
+                }
+                else
+                {
+                    /* fits entirely within litExtraBuffer, so no split is necessary */
+                    dctx->litBuffer = dctx->litExtraBuffer;
+                    dctx->litBufferEnd = dctx->litBuffer + litSize;
+                    dctx->splitLitBuffer = 0;
+                }
```

Comment:
What if you only had a `2 * WILDCOPY_OVERLENGTH` buffer for literals? I'm pretty sure my logic is correct here, but let me know if I'm missing something.

We know the maximum block size, it is stored in the dctx (`dctx->blockSize` I think). We could further refine it by taking `MIN(dctx->blockSize, remainingContentSize)` when the frame content size is set. Call this `maxBlockSize`.

Then we have two cases:

1. The output size is >= `maxBlockSize + 2 * WILDCOPY_OVERLENGTH`, then put literals ending at `oend - WILDCOPY_OVERLENGTH`. We stop `WILDCOPY_OVERLENGTH` early so we don't have to worry about overruns.
2. Else: Put the last `2 * WILDCOPY_OVERLENGTH` literals in the extra buffer, the rest go at the end.

Now, if we have a valid block, then we know that that there should always be at least `WILDCOPY_OVERLENGTH` bytes between the end of the match, and the beginning of the literal buffer. Now, we don't need to check for overrun at all, since only corrupted blocks will overrun the literals. And we don't care if we give garbage data for corrupted blocks.

We already have the check that `iLitEnd > litLimit` to call into `ZSTD_execSequenceEnd()`. When we hit that condition, in `ZSTD_execSequenceEnd()` we can see if we are in case (2). If we are, we can handle the literal special case. E.g. we set `*litPtr > litLimit` and copy from index `*litPtr - litLimit` in the extra buffer. This is UB, but we already rely on this type of UB. And doing this allows us to change absolutely nothing in the hot loop.

If we take this route, the only logic we need to change in the decoding loops is in `ZSTD_execSequenceEnd()`, which is outlined. And the rest of the hot decoding loop can be untouched.

Reply:
I'm onboard regarding the argument that, presuming input is valid, no match shall ever overlap with the literals buffer as long as we ensure a minimum distance of `WILDCOPY_OVERLENGTH `.

More difficult, I can imagine a situation where : 
- `dst` buffer is not oversized, hence it's absolutely necessary to put the last `2 * WILDCOPY_OVERLENGTH` into the internal buffer
- the input is valid
- some of the last sequences need literals from the internal buffer, while earlier sequences need literals from the `dst` buffer. And maybe one sequence needs a bit of both. 
 
In which case, collecting literals from the right place becomes a task that needs to be fulfilled. 
Your suggestion is to do it within `ZSTD_execSequenceEnd()`. 
That looks fine, we just need to be sure that it's always triggered when needed. 
Your suggestion is to rely on the existing `iLitEnd > litLimit` test,
by keeping `*litPtr` always relative to `litLimit` within `dst` buffer, 
even when the literals to collect are effectively in the internal extra-buffer.

It looks correct reasoning. I believe it could likely work.

We could remove the UB by using an index in the literals buffer rather a pointer, though it adds an extra variable, so we would need to test it does not impact register allocation too much.

The remaining question is more if we attempt this improvement as part of this PR, or as part of a subsequent PR. Let's discuss this topic depending on @binhdvo load.

Reply:
This may be separable into two parts, with one being a lower hanging fruit.  We could leave the rearrangement of the decoding logic for a later change, but in the interim simply using this more forgiving logic for deciding to use the first branch should allow us to get off the bad path with greater frequency and reduce the need for larger litExtraBuffer to handle our current bad cases.  I will experiment further.